spaceship/lib/spaceship/client.rb

```diff
@@ -679,26 +694,61 @@ def store_csrf_tokens(response)
       end
     end
 
-    def log_request(method, url, params)
+    def log_request(method, url, params, headers = nil, &block)
+      url ||= extract_key_from_block('url', &block)
+      body = extract_key_from_block('body', &block)
+      if body
+        body = JSON.parse(body)
+        body['password'] = '***' if body['password']
+      end
       params_to_log = Hash(params).dup # to also work with nil
       params_to_log.delete(:accountPassword) # Dev Portal
       params_to_log.delete(:theAccountPW) # iTC
       params_to_log = params_to_log.collect do |key, value|
         "{#{key}: #{value}}"
       end
-      logger.info(">> #{method.upcase}: #{url} #{params_to_log.join(', ')}")
+      logger.info(">> #{method.upcase} #{url}: #{body.to_json}#{params_to_log.join(', ')}")
     end
 
-    def log_response(method, url, response)
+    def log_response(method, url, response, headers = nil, &block)
+      url ||= extract_key_from_block('url', &block)
       body = response.body.kind_of?(String) ? response.body.force_encoding(Encoding::UTF_8) : response.body
-      logger.debug("<< #{method.upcase}: #{url}: #{body}")
+      resp_hash = response.to_hash
+      logger.debug("<< #{method.upcase} #{url}: #{resp_hash[:status]} #{body}")
+    end
+
+    def extract_key_from_block(key, &block)
```

Comment:
This code is terrible. Really. How do I do this (the method gets a Proc in `&block`, I want to get the values from `.body`, `.headers` and `.url()`) properly?